packages/react-native-gesture-handler/shared/RNGHRuntimeDecorator.cpp

```diff
@@ -0,0 +1,124 @@
+using namespace facebook::react;
+
+void RNGHRuntimeDecorator::installRNRuntimeBindings(
+    jsi::Runtime &rnRuntime,
+    std::function<void(int, int)> &&setGestureState) {
+  const auto isViewFlatteningDisabled = jsi::Function::createFromHostFunction(
+      rnRuntime,
+      jsi::PropNameID::forAscii(rnRuntime, "_isViewFlatteningDisabled"),
+      1,
+      [](jsi::Runtime &runtime,
+         const jsi::Value &,
+         const jsi::Value *args,
+         size_t argumentCount) -> jsi::Value {
+        if (!args[0].isObject()) {
+          return jsi::Value::null();
+        }
+
+        const auto shadowNode = shadowNodeFromValue(runtime, args[0]);
+
+#ifndef ANDROID
+        if (dynamic_pointer_cast<const ParagraphShadowNode>(shadowNode)) {
+          return jsi::Value(true);
+        }
+
+        if (dynamic_pointer_cast<const TextShadowNode>(shadowNode)) {
+          return jsi::Value(true);
+        }
+#endif
+
+        const auto isFormsStackingContext = shadowNode->getTraits().check(
+            ShadowNodeTraits::FormsStackingContext);
+
+        // This is done using component names instead of type checking because
+        // of duplicate symbols for RN types, which prevent RTTI from working.
+        const auto &componentName = shadowNode->getComponentName();
+        const auto isTextOrParagraphComponent =
+            strcmp(componentName, "Paragraph") == 0 ||
+            strcmp(componentName, "Text") == 0;
+
+        return jsi::Value(isFormsStackingContext || isTextOrParagraphComponent);
+      });
+
+  rnRuntime.global().setProperty(
+      rnRuntime,
+      "_isViewFlatteningDisabled",
+      std::move(isViewFlatteningDisabled));
+
+  auto setGestureStateAsync = jsi::Function::createFromHostFunction(
+      rnRuntime,
+      jsi::PropNameID::forAscii(rnRuntime, "_setGestureStateAsync"),
+      2,
+      [setGestureState](
+          jsi::Runtime &rt,
+          const jsi::Value &,
+          const jsi::Value *args,
+          size_t argumentCount) -> jsi::Value {
+        if (argumentCount == 2) {
+          const auto handlerTag = static_cast<int>(args[0].asNumber());
+          const auto state = static_cast<int>(args[1].asNumber());
+
+          setGestureState(handlerTag, state);
+        }
+        return jsi::Value::undefined();
+      });
+
+  rnRuntime.global().setProperty(
+      rnRuntime, "_setGestureStateAsync", std::move(setGestureStateAsync));
+}
+
+bool RNGHRuntimeDecorator::installUIRuntimeBindings(
+    jsi::Runtime &rnRuntime,
+    std::function<void(int, int)> &&setGestureState) {
+  const auto runtimeHolder =
+      rnRuntime.global().getProperty(rnRuntime, "_WORKLET_RUNTIME");
+
+  if (runtimeHolder.isUndefined()) {
+    return false;
+  }
+
+  const auto arrayBufferValue =
+      runtimeHolder.getObject(rnRuntime).getArrayBuffer(rnRuntime).data(
+          rnRuntime);
+  const auto uiRuntimeAddress =
+      reinterpret_cast<uintptr_t *>(&arrayBufferValue[0]);
+  jsi::Runtime &uiRuntime =
```

Comment:
Oh ok, so the above is just a bool specifying that there will be a buffer with a pointer to memory with uiRuntime allocated there?

Reply:
I'm not sure what you mean by "the above".

`runtimeHolder` is a jsi value holding an array buffer. `arrayBufferValue` is the data held inside the buffer (the 8 bytes of the pointer to the ui runtime). `uiRuntimeAddress` is a pointer to the first byte of the buffer value, casted to `uintptr_t` which is an integer the size of a pointer. `uiRuntime` is a reference to the ui runtime which we get by casting the `uiRuntimeAddress` to `jsi::Runtime *` and dereferencing it.

Reply:
Ok got it, thanks